reqwest-retry/src/middleware.rs

```diff
@@ -178,11 +178,24 @@ where
 
                         n_past_retries += 1;
                         continue;
-                    } else {
-                        result
                     }
                 }
-                Some(_) | None => result,
+                Some(_) | None => {}
+            };
+
+            // Report whether we failed with or without retries.
+            break if n_past_retries > 0 {
+                result.map_err(|err| {
+                    Error::Middleware(
+                        RetryError::WithRetries {
+                            retries: n_past_retries,
+                            err,
+                        }
+                        .into(),
+                    )
+                })
+            } else {
+                result.map_err(|err| Error::Middleware(RetryError::Error(err).into()))
```

Comment:
this is also not a middleware error

Reply:
It's the most common pattern in rust to have one error type per crate (or sometimes per top level module, like `serde::de::Error` and `serde::ser::Error`). Crates higher up in the dependency tree then have an enum where most variants are wrappers around the errors of crates they depend on. In reqwest-middleware, we want to allow a dynamic middleware stack, so neither do we know the list of error type statically, nor can we declare it at compile time, so we have to use a `dyn Error` such as anyhow provides. The current `request_middleware::Error` is insufficient when adding context as we do here, it doesn't properly model that the actual `reqwest::Error` instance may be lower in the stack. I don't really want to change that type to just a `anyhow::Error` wrapper with a method that finds you a `reqwest::Error` in the chain, if any, to avoid too much churn.

Reply:
the change you are making here, where normal `reqwest::Error` from the crate, is classified as `reqwest_middleware::Error::middleware` instead of `reqwest_middleware::Error::request`, is going to break every single existing user of this crate who attempts to actually inspect and handle these errors.

Reply:
it almost completely defeats the point of having `request_middleware::Error::reqwest`, which after this change will serve to only confuse users

Reply:
imo, these changelog entries are not sufficient, you should indicate a semver breaking change

Reply:
The current interpretation of the error enum assumes two cases:
* There was an internal error in the middleware
* There was a `reqwest::Error` we forward
But there exists a third class of errors, those where we do have a `reqwest::Error`, but also additional context or a middle error with a `reqwest::Error` behind it. The current enum represent this kind of errors, and there exists no simple rust pattern for matching on them.

Reply:
Good point, i've bumped `reqwest-retry` to 0.6.0, let's see what the maintainers say about `reqwest-middleware`.